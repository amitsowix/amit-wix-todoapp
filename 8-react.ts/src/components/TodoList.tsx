import React, {FC, useEffect, useState} from 'react';
import {createUseStyles} from 'react-jss'
import {Classes} from 'jss';
import Button from './Button';
import Input from './Input';
import TodoItem from './TodoItem'
import ServerAPI from '../helpers/serverapi'
import { TodoItemDTO, TodoListDTO } from '../../../common/interfaces';
import { v4 as uuidv4 } from 'uuid';


const TodoList: FC = (): JSX.Element => {

    const [todoListDTO, setTodoListDTO] = useState<TodoListDTO>({});
    const [isInEditMode, setEditMode] = useState<boolean>(false);

    const classes: Classes = useStyles();

    const getTodoList: () => void = async () => {
        const todoList: TodoListDTO | undefined = await ServerAPI.getTodoItemsList();
        setTodoListDTO(todoList || {});
    }

    const addNewItem = (text: string): void => {
        const id = uuidv4();
        const newItem: TodoItemDTO = {
            text,
            id,
            isChecked: false
        }
        const newList = {...todoListDTO};
        newList[id] = newItem;
        setTodoListDTO(newList);
        ServerAPI.addTodoItem(newItem);
    }

    const editTodoItem = (updatedItem: TodoItemDTO): void => {
        const newList = {...todoListDTO};
        newList[updatedItem.id] = updatedItem;
        setTodoListDTO(newList);
        ServerAPI.editTodoItem(updatedItem.id, updatedItem);
    }

    const deleteTodoItem = (id: string): void => {
        const newList = {...todoListDTO};
        delete newList[id];
        setTodoListDTO(newList);
        ServerAPI.deleteTodoItem(id);
    }
    
    useEffect(getTodoList, []);

    return (
        <div className = {classes.appBodyContainer}>
            <div className={classes.newItemContainer}>
                <Button onClick={()=>setEditMode(!isInEditMode)}/>
                {isInEditMode ? <Input onInputSubmit={addNewItem} /> : null}
            </div>
            <ul className={classes.listContainer}>
                {Object.values(todoListDTO).map((item: TodoItemDTO) : JSX.Element => {
                    return <TodoItem 
                        key={item.id} id={item.id} 
                        text={item.text} 
                        isChecked={item.isChecked} 
                        editTodoItem={editTodoItem} 
                        deleteTodoItem={deleteTodoItem}
                    />
                })}
            </ul>
        </div>
    )            
}

const useStyles : () => Classes = createUseStyles({
    appBodyContainer: {
        padding: "50px 70px"
    },
    newItemContainer: {
        display: "flex",
        alignItems: "flex-end"
    },
    icon: {
        height: "30px",
        marginLeft: "10px",
        cursor: "pointer"
    },
    listContainer: { 
        height: "400px" 
    }
});

export default TodoList;