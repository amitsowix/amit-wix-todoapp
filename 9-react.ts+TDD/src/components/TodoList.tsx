import React, {FC, useEffect, useState} from 'react';
import {createUseStyles} from 'react-jss'
import {Classes} from 'jss';
import Button from './Button';
import Input from './Input';
import TodoItem from './TodoItem'
import ServerAPI from '../helpers/serverapi'
import { TodoItemDTO, TodoListDTO } from '../../../common/interfaces';
import { v4 as uuidv4 } from 'uuid';


const TodoList: FC = (): JSX.Element => {

    const [todoList, setTodoList] = useState<TodoListDTO>({});
    const [isInEditMode, setEditMode] = useState<boolean>(false);
    const [showError, setShowError] = useState<boolean>(false);

    const classes: Classes = useStyles();

    const getTodoList: () => void = async () => {
        try {
            const todoList: TodoListDTO = await ServerAPI.getTodoItemsList();
            setTodoList(todoList);
            setShowError(false);
        }
        catch (e){
            setShowError(true);
        }
    }

    const addNewItem = async (text: string) => {
        const id = uuidv4();
        const newItem: TodoItemDTO = {
            text,
            id,
            isChecked: false
        }
        const newList = {...todoList};
        newList[id] = newItem;
        setTodoList(newList);
        try {
            await ServerAPI.addTodoItem(newItem);
            setShowError(false);
        }
        catch (e) {
            setShowError(true);
            const oldList = {...newList};
            delete oldList[id];
            setTodoList(oldList);
        }
    }

    const editTodoItem =  async (updatedItem: TodoItemDTO) => {
        const oldItem = {...todoList[updatedItem.id]}
        const newList = {...todoList};
        newList[updatedItem.id] = updatedItem;
        setTodoList(newList);
        try {
            await ServerAPI.editTodoItem(updatedItem.id, updatedItem);
            setShowError(false);
        }
        catch (e){
            setShowError(true);
            const oldList = {...newList};
            oldList[updatedItem.id] = oldItem
            setTodoList(oldList);
        }
    }

    const deleteTodoItem = async (id: string) => {
        const newList = {...todoList};
        const deletedItem = newList[id];
        delete newList[id];
        setTodoList(newList);
        try {
            await ServerAPI.deleteTodoItem(id);
            setShowError(false);
        }
        catch (e){
            setShowError(true);
            const oldList = {...newList};
            oldList[id] = deletedItem;
            setTodoList(oldList);
        }
    }
    
    useEffect(getTodoList, []);

    return (
        <div className = {classes.appBodyContainer}>
            <div className={classes.newItemContainer}>
                <Button onClick={()=>setEditMode(!isInEditMode)}/>
                {isInEditMode ? <Input name={'add-new-item'} onInputSubmit={addNewItem} /> : null}
            </div>
            {showError ? <div className={classes.error}>Error Connecting to DB</div> : null}
            <ul data-hook={'list-container'} className={classes.listContainer}>
                {Object.values(todoList).map((item: TodoItemDTO) : JSX.Element => {
                    return <TodoItem 
                        key={item.id} id={item.id} 
                        text={item.text} 
                        isChecked={item.isChecked} 
                        editTodoItem={editTodoItem} 
                        deleteTodoItem={deleteTodoItem}
                    />
                })}
            </ul>
        </div>
    )            
}

const useStyles : () => Classes = createUseStyles({
    appBodyContainer: {
        padding: "50px 70px"
    },
    newItemContainer: {
        display: "flex",
        alignItems: "flex-end"
    },
    icon: {
        height: "30px",
        marginLeft: "10px",
        cursor: "pointer"
    },
    listContainer: { 
        height: "400px" 
    },
    error: {
        color: 'red',
    }
});

export default TodoList;