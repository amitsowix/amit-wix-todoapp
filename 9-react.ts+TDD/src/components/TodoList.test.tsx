import React from 'react';
import renderer, {ReactTestRenderer, ReactTestRendererJSON} from 'react-test-renderer';
import {screen, render} from '@testing-library/react';
import TodoList from './TodoList';
import {mount, ReactWrapper} from 'enzyme';


const driver = {
    when: {
        mount: (): ReactTestRenderer => {
            const component: ReactTestRenderer = renderer.create(
                <TodoList/>
            )            
            return component;
        },
        enzymeMount: (): ReactWrapper => {
            const wrapper = mount(<TodoList/>);
            return wrapper;
        },
        click: (element: ReactWrapper): void => {
            element.simulate('click');
        }
    },
    get: {
      addNewItemButton: (wrapper: ReactWrapper) => wrapper.find('div').at(2),
      submitButton: (wrapper: ReactWrapper) => wrapper.find('div[data-hook="submit-button"]'),
      input: (wrapper: ReactWrapper) => wrapper.find('input').at(0),
    }
}


describe('Test Render', () => {
    it('should render correctly', () => {

        const component: ReactTestRenderer = driver.when.mount();

        const tree: ReactTestRendererJSON | ReactTestRendererJSON[] | null = component.toJSON();

        expect(tree).toMatchSnapshot();
    })

    it('should be of type div', () => {

        const component: ReactTestRenderer = driver.when.mount();

        const tree: ReactTestRendererJSON | ReactTestRendererJSON[] | null = component.toJSON();
        if (tree && !Array.isArray(tree)){
            expect(tree.type).toMatch('div');
        }
    })

    it('should not show the add new item input', () => {

        const wrapper = driver.when.enzymeMount();

        expect(driver.get.input(wrapper).exists()).toBeFalsy();
        expect(driver.get.submitButton(wrapper).exists()).toBeFalsy();
    })
})


describe('Input show/hide', () => {
    it('should Show/hide input on click', () => {

        const wrapper: ReactWrapper = driver.when.enzymeMount();

        const button: ReactWrapper = driver.get.addNewItemButton(wrapper);

        driver.when.click(button);

        const input: ReactWrapper = driver.get.input(wrapper);
        const submitButton: ReactWrapper = driver.get.submitButton(wrapper);

        expect(input.exists()).toBeTruthy();
        expect(submitButton.exists()).toBeTruthy();

        driver.when.click(button);

        expect(input).not.toBeInTheDocument;
        expect(submitButton).not.toBeInTheDocument;
    })
})

