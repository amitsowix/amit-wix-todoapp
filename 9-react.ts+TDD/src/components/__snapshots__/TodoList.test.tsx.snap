// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Render should render correctly 1`] = `
<div
  className="appBodyContainer-0-2-1"
>
  <div
    className="newItemContainer-0-2-2"
  >
    <div
      className="newItemButton-0-2-6"
      data-hook="new-item-button"
      onClick={[Function]}
    >
      <img
        alt=""
        className="addIcon-0-2-7"
        src="test-file-stub"
      />
      <div>
        New Item
      </div>
    </div>
  </div>
  <ul
    className="listContainer-0-2-4"
    data-hook="list-container"
  />
</div>
`;
